### DISCUSSION TITLE
Offline map

### DISCUSSION THREAD

**Discussion by wipfli** [protomaps/PMTiles#94] [Q&A] [answered]
What is the best way to use PMTiles for an offline map?

Ideally, I would like to have a single html file with all the javascript embedded and maybe the pmtiles file could be base64 encoded and stored as a string in javascript or so. Do you think that is possible?

**Comment by bdon** [comment, marked as answer]
Run a local HTTP server.

The other option is to use the `file://` protocol, but this requires user interaction to access the filesystem for security reasons. 

The base64 idea could also work but seems like it would be pretty ugly after a few megabytes. You'd have to write a new class implementing the `Source` interface that is backed by a simple ArrayBuffer loaded from the base64 string.

**Reply by wipfli** [reply]
```js
  var FetchSource = class {
    constructor(url) {
      this.url = url;
      const output = "UE1U..."; // base64 encoded pmtiles file of Monaco
      function _base64ToArrayBuffer(base64) {
        var binary_string = window.atob(base64);
        var len = binary_string.length;
        var bytes = new Uint8Array(len);
        for (var i = 0; i < len; i++) {
          bytes[i] = binary_string.charCodeAt(i);
        }
        return bytes.buffer;
      }
      this.buffer = _base64ToArrayBuffer(output);
    }
    getKey() {
      return this.url;
    }
    getBytes(offset, length, signal) {
      return __async(this, null, function* () {
        return {
          data: this.buffer.slice(offset, offset + length)
        };
      });
    }
  };
```

**Reply by wipfli** [reply]
Was an ugly quicky hack into the bundled pmtiles javascript index.js file, and actually worked. I tried it with the Monaco .pmtiles openmaptiles file generated with Planetiler. This file is roughly 350 kB.

**Reply by wipfli** [reply]
Somehow it would be really cool if we could do stuff like this:

```json
"sources": {
  "mytiles": {
    "type": "vector",
    "url": "pmtiles://data:application/octet-stream;base64,UE1UaWxl..."
  }
}
```

**Reply by wipfli** [reply]
I mean you are right @bdon that this becomes ugly past a few megabytes, but I would find it fun...

**Reply by wipfli** [reply]
Here is a demo: https://wipfli.github.io/swiss-map/base64/#18/43.733407/7.419744

**Reply by bdon** [reply]
@wipfli awesome!

I think it's really cool but adding another code path for base64-encoded PMTiles means another thing to maintain for a niche use case. If your data is small enough to fit inline in the web page I think non-tiled GeoJSON is always a better choice for that. I'm having a hard time thinking of practical examples where this would be a superior choice.

**Reply by wipfli** [reply]
You are right, GeoJSON seems better for offline maps. Thanks for the hint...

**Reply by do-me** [reply]
@wipfli could you share that code/example again please? The link is broken. Would like to play with it too ;D

But I think I agree with bdon, that it doesn't make much sense as it always duplicates data (for the lowest zoom levels). However, what if we kept only the full data (without building any pyramids/ lower zoom levels) and just adapted the style.json for generalization to leave out certain features at certain zoom levels? 

Also, are you aware of any minimal alternatives for "hard-coding" a basemap into a minimal index.html? This whole minimalist-low-file-size mapping idea really got me hooked!

**Reply by wipfli** [reply]
The code is here: https://github.com/wipfli/swiss-map/tree/8269fb403f29e59a99b25b3d9e105df65e816b62/base64
